File: day-1/go/one.go
package main

import (
	"aoc24/utils"
	"fmt"
	"slices"
	"strconv"
	"strings"
)

func getLists() ([]int, []int) {
	lines := utils.ReadLines("../inputs/input-one.txt")

	list1 := []int{}
	list2 := []int{}

	for _, line := range lines {
		result := strings.Split(line, "   ")

		first, err := strconv.Atoi(result[0])
		if err != nil {
			panic(err)
		}

		list1 = append(list1, first)

		second, err := strconv.Atoi(result[1])
		if err != nil {
			panic(err)
		}

		list2 = append(list2, second)
	}

	return list1, list2
}

func sumDistances(list1 []int, list2 []int) int {
	sum := 0

	for i := range list1 {
		sum += utils.Abs(list1[i] - list2[i])
	}

	return sum
}

func main() {
	list1, list2 := getLists()

	slices.Sort(list1)
	slices.Sort(list2)

	fmt.Println(sumDistances(list1, list2)) // 1197984
}

File: day1/one.ts
import { readInputByLines } from '../utils/index.ts';

type Lists = [number[], number[]];

const getListsFromInput = (): Lists => {
	const lists: Lists = [[], []];

	readInputByLines('../inputs/input-one.txt')
		.map((line) => line.split('   ').map(Number))
		.forEach(([first, second]) => {
			lists[0].push(first);
			lists[1].push(second);
		});

	return lists;
};

const sumDistances = (list1: number[], list2: number[]): number => {
	return list1.reduce((sum, item, idx) => sum + Math.abs(item - list2[idx]), 0);
};

const main = (): number => {
	const [list1, list2] = getListsFromInput();

	list1.sort();
	list2.sort();

	return sumDistances(list1, list2);
};

console.log('result:', main()); // success: 1197984
